app/src/org/commcare/utils/Permissions.java

```diff
@@ -88,38 +107,147 @@ public static boolean shouldShowPermissionRationale(AppCompatActivity activity,
+            if (which == AlertDialog.BUTTON_POSITIVE) {
+                switch(specialPerm) {
+                    case Manifest.permission.SCHEDULE_EXACT_ALARM:
+                        startAlarmsAndRemindersActivity(activity);
+                        break;
+                    default:
+                        throw new RuntimeException("Invalid special permission " + specialPerm);
+                }
+            }
+            else
+                communicateFeatureDegradation(activity, specialPerm);
+            dialog.dismiss();
+        };
+        d.setPositiveButton(Localization.get("permission.acquire.dialog.allow"), listener);
+        d.setNegativeButton(Localization.get("permission.acquire.dialog.deny"), listener);
+        d.showNonPersistentDialog();
+    }
+
+    public static void communicateFeatureDegradation(AppCompatActivity activity, String specialPerm) {
+        // Any other measures should be registered in the switch below
+        switch(specialPerm){
+            case Manifest.permission.SCHEDULE_EXACT_ALARM:
+                // Don't check this permission again, the user will have to go to Apps -> Special access
+                // to grant it
+                HiddenPreferences.setCheckScheduleExactAlarmPermission(false);
+                break;
+            default:
+                throw new RuntimeException("Invalid special permission " + specialPerm);
+        }
+        StandardAlertDialog d = StandardAlertDialog.getBasicAlertDialog(activity,
+                Localization.get(getSpecialPermissionDialogTitle(specialPerm)),
+                Localization.get(getSpecialPermissionDenialDialogMessage(specialPerm)),null);
+        d.showNonPersistentDialog();
+    }
+
+    private static String getSpecialPermissionDenialDialogMessage(String specialPerm) {
+        switch(specialPerm){
+            case Manifest.permission.SCHEDULE_EXACT_ALARM:
+                return "permission.schedule.exact.alarm.denied";
+            default:
+                throw new RuntimeException("Invalid special permission " + specialPerm);
+        }
+    }
+
+    private static String getSpecialPermissionRequestDialogMessage(String specialPerm) {
+        switch(specialPerm){
+            case Manifest.permission.SCHEDULE_EXACT_ALARM:
+                return "permission.schedule.exact.alarm.message";
+            default:
+                throw new RuntimeException("Invalid special permission " + specialPerm);
+        }
+    }
+
+    private static String getSpecialPermissionDialogTitle(String specialPerm) {
+        switch(specialPerm){
+            case Manifest.permission.SCHEDULE_EXACT_ALARM:
+                return "permission.schedule.exact.alarm.title";
+            default:
+                throw new RuntimeException("Invalid special permission " + specialPerm);
+        }
+    }
+
+    // TODO: In addition to scrolling to where CommCare app is, we should highlight the item
```

Comment:
Does the page this intent send us to not CC specific and user need to locate CC from a list of apps ?